bigtop-packages/src/charm/hadoop/layer-hadoop-plugin/reactive/apache_bigtop_plugin.py

```diff
@@ -0,0 +1,149 @@
+@when_not('namenode.joined')
+def blocked(principal):
+    hookenv.status_set('blocked', 'missing required namenode relation')
+
+
+@when('bigtop.available', 'hadoop-plugin.joined', 'namenode.joined')
+@when_not('apache-bigtop-plugin.hdfs.installed')
+def install_hadoop_client_hdfs(principal, namenode):
+    """Install if the namenode has sent its FQDN.
+
+    We only need the namenode FQDN to perform the plugin install, so poll for
+    namenodes() data whenever we have a namenode relation. This allows us to
+    install asap, even if 'namenode.ready' is not set yet.
+    """
+    if namenode.namenodes():
+        hookenv.status_set('maintenance', 'installing plugin (hdfs)')
+        nn_host = namenode.namenodes()[0]
+        bigtop = Bigtop()
+        hosts = {'namenode': nn_host}
+        bigtop.render_site_yaml(hosts=hosts, roles='hadoop-client')
+        bigtop.trigger_puppet()
+        set_state('apache-bigtop-plugin.hdfs.installed')
+        hookenv.status_set('maintenance', 'plugin (hdfs) installed')
+    else:
+        hookenv.status_set('waiting', 'waiting for namenode fqdn')
+
+
+@when('apache-bigtop-plugin.hdfs.installed')
+@when('hadoop-plugin.joined', 'namenode.joined')
+@when_not('namenode.ready')
+def send_nn_spec(principal, namenode):
+    """Send our plugin spec so the namenode can become ready."""
+    bigtop = Bigtop()
+    # Send plugin spec (must match NN spec for 'namenode.ready' to be set)
+    namenode.set_local_spec(bigtop.spec())
+
+
+@when('apache-bigtop-plugin.hdfs.installed')
+@when('hadoop-plugin.joined', 'namenode.ready')
+@when_not('apache-bigtop-plugin.hdfs.ready')
+def send_principal_hdfs_info(principal, namenode):
+    """Send HDFS data when the namenode becomes ready."""
+    principal.set_installed(get_hadoop_version())
+    principal.set_hdfs_ready(namenode.namenodes(), namenode.port())
+    set_state('apache-bigtop-plugin.hdfs.ready')
+
+
+@when('apache-bigtop-plugin.hdfs.ready')
+@when('hadoop-plugin.joined')
+@when_not('namenode.ready')
+def clear_hdfs_ready(principal):
+    principal.clear_hdfs_ready()
+    remove_state('apache-bigtop-plugin.hdfs.ready')
+    remove_state('apache-bigtop-plugin.hdfs.installed')
+
+
+@when('bigtop.available', 'hadoop-plugin.joined', 'namenode.joined', 'resourcemanager.joined')
+@when_not('apache-bigtop-plugin.yarn.installed')
+def install_hadoop_client_yarn(principal, namenode, resourcemanager):
+    if namenode.namenodes() and resourcemanager.resourcemanagers():
+        hookenv.status_set('maintenance', 'installing plugin (yarn)')
+        nn_host = namenode.namenodes()[0]
+        rm_host = resourcemanager.resourcemanagers()[0]
+        bigtop = Bigtop()
+        hosts = {'namenode': nn_host, 'resourcemanager': rm_host}
+        bigtop.render_site_yaml(hosts=hosts, roles='hadoop-client')
+        bigtop.trigger_puppet()
+        set_state('apache-bigtop-plugin.yarn.installed')
+        hookenv.status_set('active', 'ready (HDFS & YARN)')
+    else:
```

Comment:
How often will this trigger, while it is waiting? Will it trigger as many times per second as the CPU can run it, or do we have a built in timer that will prevent it from firing again too soon?